src/libcollections/vec.rs

```diff
@@ -1748,14 +1741,77 @@ impl<T> Drop for IntoIter<T> {
     }
 }
 
-/// An iterator that drains a vector.
+/// A trait for draining a vector.
+///
+/// See the documentation of `Vec::drain`.
+pub trait DrainRange {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T>;
+}
+
+impl DrainRange for Range<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        assert!(self.start <= self.end, "Range not increasing");
+        assert!(self.end <= vec.len(), "Range out of bounds");
+        unsafe {
+            let tail = vec.len() - self.end;
+            vec.set_len(tail + self.start);
+            let ptr = vec.as_ptr();
+            let (start, end) = if mem::size_of::<T>() == 0 {
+                // make sure that start is not null
+                ((self.start + 1) as *const T, (self.end + 1) as *const T)
+            } else {
+                (ptr.offset(self.start as isize), ptr.offset(self.end as isize))
+            };
+            Drain {
+                tail:  tail,
+                start: start,
+                end:   end,
+                left:  start,
+                right: end,
+                marker1: PhantomData,
+                marker2: PhantomData,
+            }
+        }
+    }
+}
+
+impl DrainRange for RangeFrom<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        assert!(self.start <= vec.len(), "Range out of bounds");
+        (self.start..vec.len()).drain(vec)
+    }
+}
+
+impl DrainRange for RangeTo<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (0..self.end).drain(vec)
+    }
+}
+
+impl DrainRange for RangeFull {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (0..vec.len()).drain(vec)
+    }
+}
+
+impl DrainRange for usize {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (*self..*self+1).drain(vec)
+    }
+}
+
+/// An iterator that drains part of a vector.
 #[unsafe_no_drop_flag]
 #[unstable(feature = "collections",
            reason = "recently added as part of collections reform 2")]
-pub struct Drain<'a, T:'a> {
-    ptr: *const T,
-    end: *const T,
-    marker: PhantomData<&'a T>,
+pub struct Drain<'a, T: 'a> {
+    tail: usize,
+    start: *const T,
+    end:   *const T,
+    left:  *const T,
+    right: *const T,
+    marker1: PhantomData<&'a ()>,
+    marker2: PhantomData<T>,
```

Comment:
If I were going to write `Drain` in pure safe code, it would look something like:

```rust
pub struct Drain<'a, T:'a> {
    start: uint,
    len: uint,
    victim: &'a Vec<T>
}
```

then on each call to `next` I could call `remove` on the item at `self.start` and decrement `self.len`. When dropping I would keep calling `next` `self.len` times. This should be roughly equivalent to the current behavior, though the modifications to the vector happen at different times, but I believe that is unobservable.

Therefore, I believe the marker you want is `PhantomData<&'a mut Vec<T>>`. 

It is reasonable to want to put `PhantomData<T>`, because the destructor will drop instances of `T`, but not necessary -- the dropck will already be imposing conservative rules on `Drain` because it contains borrowed data. Basically, if you model what you would have written in safe code, you should be ok (unless the language itself is flawed).

cc @pnkfelix